mod-source-record-storage-server/src/main/java/org/folio/services/handlers/actions/AbstractDeleteEventHandler.java

```diff
@@ -0,0 +1,103 @@
+import org.folio.processing.events.services.handler.EventHandler;
+import org.folio.processing.exceptions.EventProcessingException;
+import org.folio.rest.jaxrs.model.ProfileSnapshotWrapper;
+import org.folio.rest.jaxrs.model.Record;
+import org.folio.services.RecordService;
+import org.folio.services.util.TypeConnection;
+
+import java.util.concurrent.CompletableFuture;
+
+import static java.util.Objects.isNull;
+import static org.apache.commons.lang3.StringUtils.isBlank;
+import static org.folio.ActionProfile.Action.DELETE;
+import static org.folio.rest.jaxrs.model.ProfileSnapshotWrapper.ContentType.ACTION_PROFILE;
+
+/**
+ * The abstraction handles the DELETE action
+ * The handler:
+ * 1. Validates the event payload
+ * 2. Retrieves a matched record from the context
+ * 3. Deletes the record passing a matched record and tenant id
+ * 4. If successfully deleted - replaces a record on FOLIO record id in the context,
+ * else completes exceptionally and loggs a cause
+ */
+public abstract class AbstractDeleteEventHandler implements EventHandler {
+  private static final Logger LOG = LogManager.getLogger();
+  private static final String PAYLOAD_HAS_NO_DATA_MSG = "Failed to handle event payload, cause event payload context does not contain required data to modify MARC record";
+  private static final String ERROR_WHILE_DELETING_MSG = "Error while deleting MARC record, record is not found";
+  protected final TypeConnection typeConnection;
+  protected final RecordService recordService;
+
+  public AbstractDeleteEventHandler(RecordService recordService, TypeConnection typeConnection) {
+    this.recordService = recordService;
+    this.typeConnection = typeConnection;
+  }
+
+  @Override
+  public CompletableFuture<DataImportEventPayload> handle(DataImportEventPayload payload) {
+    CompletableFuture<DataImportEventPayload> future = new CompletableFuture<>();
+    var payloadContext = payload.getContext();
+    if (isNull(payloadContext) || isBlank(payloadContext.get(getRecordKey()))) {
+      completeExceptionally(future, new EventProcessingException(PAYLOAD_HAS_NO_DATA_MSG));
+      return future;
+    } else {
+      payload.getEventsChain().add(payload.getEventType());
+      try {
+        handlePayload(payload, future);
+      } catch (Throwable exception) {
+        completeExceptionally(future, exception);
+      }
+    }
+    return future;
+  }
+
+  /* Handles DELETE action  */
+  private void handlePayload(DataImportEventPayload payload, CompletableFuture<DataImportEventPayload> future) {
+    var record = Json.decodeValue(payload.getContext().get(getRecordKey()), Record.class);
+    recordService.updateRecord(record.withState(Record.State.DELETED), payload.getTenant())
+      .onSuccess(ar -> {
+        payload.setEventType(getNextEventType());
+        payload.getContext().remove(getRecordKey());
+        payload.getContext().put(getExternalRecordIdKey(), record.getExternalIdsHolder().getAuthorityId());
```

Comment:
`record.getExternalIdsHolder().getAuthorityId()` should be extracted to abstract method and overridden in `MarcAuthorityDeleteEventHandler`. This is specific to authority flow.